### PR TITLE
feat: add ACL on custom link

### PR DESCRIPTION
**Pull request by gcusnieux**
fix gravitee-io/issues#4563

### CODE REVIEW

File: gravitee-rest-api-management/gravitee-rest-api-management-rest/src/test/java/io/gravitee/rest/api/management/rest/resource/ApiPagesResourceNotAdminTest.java [hunk 1]
```diff
@@ -95,17 +93,15 @@ public void shouldNotGetPrivateApiPublishedPage() {
         pageMock.setPublished(true);
         pageMock.setName(PAGE_NAME);
         doReturn(pageMock).when(pageService).findById(PAGE_NAME, null);
-        doReturn(false).when(pageService).isDisplayable(apiMock, pageMock.isPublished(), USER_NAME);
         doReturn(true).when(roleService).hasPermission(any(), eq(ApiPermission.DOCUMENTATION), eq(new RolePermissionAction[]{RolePermissionAction.READ}));
-        when(groupService.isUserAuthorizedToAccessApiData(any(), any(), any())).thenReturn(Boolean.FALSE);
+//        when(groupService.isUserAuthorizedToAccessApiData(any(), any(), any())).thenReturn(Boolean.FALSE);
```
Comment (ytvnr) [inline gravitee-rest-api-management/gravitee-rest-api-management-rest/src/test/java/io/gravitee/rest/api/management/rest/resource/ApiPagesResourceNotAdminTest.java:97]: can you delete it if useless ?

File: gravitee-rest-api-model/src/main/java/io/gravitee/rest/api/model/PageEntity.java [hunk 2]
```diff
@@ -44,8 +41,8 @@
     private boolean homepage;
     private String parentId;
     private String parentPath;
-    @JsonProperty("excluded_groups")
-    private List<String> excludedGroups;
+    private boolean excludedAccessControls;
```
Comment (ytvnr) [inline gravitee-rest-api-model/src/main/java/io/gravitee/rest/api/model/PageEntity.java:44]:
```suggestion
    @JsonProperty("excluded_access_controls")
    private boolean excludedAccessControls;
```

File: gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/test/java/io/gravitee/rest/api/portal/rest/resource/ApiPagesResourceTest.java [hunk 3]
```diff
@@ -52,25 +46,31 @@
 
     private static final String API = "my-api";
 
+
+
```
Comment (ytvnr) [inline gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/test/java/io/gravitee/rest/api/portal/rest/resource/ApiPagesResourceTest.java:50]: useless

File: gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/test/java/io/gravitee/rest/api/portal/rest/resource/ApiPagesResourceTest.java [hunk 4]
```diff
@@ -52,25 +46,31 @@
...
         ApiEntity mockApi = new ApiEntity();
         mockApi.setId(API);
         doReturn(mockApi).when(apiService).findById(API);
-        
-        Set<ApiEntity> mockApis = new HashSet<>(Arrays.asList(mockApi));
-        doReturn(mockApis).when(apiService).findPublishedByUser(any(), argThat(q -> singletonList(API).equals(q.getIds())));
 
+//        Set<ApiEntity> mockApis = new HashSet<>(Arrays.asList(mockApi));
+//        doReturn(mockApis).when(apiService).findPublishedByUser(any(), argThat(q -> singletonList(API).equals(q.getIds())));
```
Comment (ytvnr) [inline gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/test/java/io/gravitee/rest/api/portal/rest/resource/ApiPagesResourceTest.java:64]: to delete

File: gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/test/java/io/gravitee/rest/api/portal/rest/resource/ApiPagesResourceTest.java [hunk 5]
```diff
@@ -115,28 +114,18 @@ public void shouldGetApiPages() {
...
-        
-        response = request.get();
-        assertEquals(OK_200, response.getStatus());
-
-        pagesResponse = response.readEntity(PagesResponse.class);
-        pages = pagesResponse.getData();
-        assertNotNull(pages);
-        assertEquals(0, pages.size());
     }
+
```
Comment (ytvnr) [inline gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/test/java/io/gravitee/rest/api/portal/rest/resource/ApiPagesResourceTest.java:130]: useless

File: gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/test/java/io/gravitee/rest/api/portal/rest/resource/PageResourceNotAuthenticatedTest.java [hunk 6]
```diff
@@ -79,26 +80,27 @@ public boolean isUserInRole(String string) {
 
     private PageEntity mockAnotherPage;
 
-    
+
```
Comment (ytvnr) [inline gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/test/java/io/gravitee/rest/api/portal/rest/resource/PageResourceNotAuthenticatedTest.java:83]: useless

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AccessControlServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    @Override
+    public boolean canAccessPageFromConsole(ApiEntity apiEntity, PageEntity pageEntity) {
+        if (canAccessPage(apiEntity, pageEntity)) {
+            return true;
+        } else {
+            return canEditDocumentation(apiEntity);
+        }
+    }
+
+    private boolean canEditDocumentation(ApiEntity api) {
```
Comment (ytvnr) [inline gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AccessControlServiceImpl.java:138]: ⚠️ I think api can be null

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AccessControlServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+        boolean canEditDocumentation = false;
+        if (isAuthenticated()) {
+            MemberEntity member = membershipService.getUserMember(MembershipReferenceType.API, api.getId(), getAuthenticatedUsername());
```
Comment (ytvnr) [inline gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AccessControlServiceImpl.java:141]: this method seems only related to membershipReferenceType.API, does it work well for platform level ?

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AccessControlServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+
+    private boolean canEditDocumentation(MemberEntity member) {
+        // if not member => not displayable
+        if (member == null) {
+            return false;
+        }
+
+        // only members which could modify a page can see an unpublished page
+        return roleService.hasPermission(member.getPermissions(), ApiPermission.DOCUMENTATION,
```
Comment (ytvnr) [inline gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/AccessControlServiceImpl.java:164]: same question, it seems we use only ApiPermission

**Review by ytvnr — COMMENTED** [review]
some comments

File: gravitee-rest-api-model/src/main/java/io/gravitee/rest/api/model/PageEntity.java (hunk 2 above)
Comment (gcusnieux) [reply gravitee-rest-api-model/src/main/java/io/gravitee/rest/api/model/PageEntity.java:44]: Why?  In this case, we should change all properties parent_id, parent_path... I'm from js world, I like camelCase

**Review by ytvnr — APPROVED** [review]

**Pull request merged**